docs_src/microservices/security/Ch-Docker-Swarm-SecuringDeviceServices.md

```diff
@@ -0,0 +1,218 @@
+
+The following command initializes a docker swarm and is to be ran on the host machine: 
+
+```sh
+sudo docker swarm init --advertise-addr <your ip address>
+```
+
+The previous command will output a token use this token in the following join command.  This joins the worker node to the cluster, to be ran your vagrant VM (worker-node):
+
+```sh
+sudo docker swarm join --token <token> <your ip manager node's address>:2377
+```
+
+Next, I will walk-through the changes we made to the docker-stack.yml file to convert the edgex compose file into a docker swarm stack file.
+
+### Setting up the docker-stack-edgex.yml file
+
+All of the following changes are already done in the examples repo.  I will just outline the necessary changes to go from a compose file to stack file.
+
+First, remove 'restart' command from compose file; 'restart' is not a valid command in docker swarm.
+
+Next, we define constraints that the edgex core services must not run on the worker node add this section of yml to the 'docker-stack-edgex.yml'.  We will do the inverse for the device-service to ensure it does run on the worker node thus ensuring it uses the overlay network to communicate with the other services.  Note that this is already done in the example directory.
+
+```yaml
+    deploy: 
+      placement: 
+        constraints: 
+          - node.hostname != worker-node
+```
+Here is the inverse of the previous yml block. This gets added to the device services in the stack file. 
+
+```yaml
+    deploy: 
+      placement: 
+        constraints: 
+          - node.hostname == worker-node
+```
+
+### Adding Host Mounted Volumes
+
+In docker swarm bind mount volumes need to be explicitly defined in the stack yml file.  Below are the fist three bind mount volume definitions.  Additionally, you must create the directories on the host before the stack file in ran.
+
+Note that, this is only an example in a production deployment you would want to use a network filesystem or create the shared volumes between containers.
+
+```yaml
+  secrets-volume:
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /tmp/edgex/secrets/
+  secrets-ca-volume:
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /tmp/edgex/secrets/ca/
+  edgex-consul:
+    driver: local
+    driver_opts:
+      o: bind
+      type: none
+      device: /tmp/edgex/secrets/edgex-consul/
+...
+```
+
+The full docker-stack file is included here:
+
+[docker-stack-edgex.yml file](https://github.com/edgexfoundry/edgex-examples/security/remote_devices/docker-swarm/docker-stack-edgex.yml)
+
+### Other changes in docker-stack-edgex.yml file 
+
+Another change we had to make in the docker-stack-edgex.yml file is to disable `IPC_LOCK` because the `cap_add` flag in vault's configuration is not supported in docker swarm.  To do this we add `SKIP_SETCAP=true` and `disable_mlock = "true"` to vault in the stack file. 
```

Comment:
Also add another description on changes for one-shot containers (like only run once and gone) via using `Restart-Policy` element in docker stack.

Reply:
good call